File: 102.binary-tree-level-order-traversal.js
/*
 * @lc app=leetcode id=102 lang=javascript
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {number[][]}
 */
var levelOrder = function(root) {
  if (!root) return []
  // O(n) time
  // O(n) space
  // store node + level
  let queue = []
  let levelOrderTraversal = []
  queue.unshift([root, 0])
  while (queue.length > 0) {
    let vals = queue.shift()
    let currNode = vals[0]
    let currLevel = vals[1]
    if (levelOrderTraversal[currLevel]) {
      levelOrderTraversal[currLevel].push(currNode.val)
    } else {
      levelOrderTraversal[currLevel] = [currNode.val]
    }
    if (currNode.left) {
      queue.push([currNode.left, currLevel+1])
    }
    if (currNode.right) {
      queue.push([currNode.right, currLevel+1])
    }
  }
  return levelOrderTraversal
};
// @lc code=end

File: cpp/102.binary-tree-level-order-traversal.cpp
/*
 * @lc app=leetcode id=102 lang=cpp
 *
 * [102] Binary Tree Level Order Traversal
 */

// @lc code=start
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution
{
public:
    struct QueueVal
    {
        TreeNode *root;
        int level;
    };
    vector<vector<int>> levelOrder(TreeNode *root)
    {
        // BFS traversal
        // keep track of levels
        // O(n) time O(n) space
        vector<vector<int>> resVect;
        if (!root)
            return resVect;
        queue<QueueVal> queuer;
        queuer.push(QueueVal{root, 0});
        while (queuer.size() > 0)
        {
            QueueVal vals = queuer.front();
            TreeNode *currNode = vals.root;
            int currLevel = vals.level;
            queuer.pop();
            // check if there exists previous vect at level
            if (currLevel == resVect.size())
            {
                vector<int> newVect = {currNode->val};
                resVect.push_back(newVect);
            }
            else
            {
                resVect[currLevel].push_back(currNode->val);
            }

            if (currNode->left)
            {
                queuer.push(QueueVal{currNode->left, currLevel + 1});
            }
            if (currNode->right)
            {
                queuer.push(QueueVal{currNode->right, currLevel + 1});
            }
        }
        return resVect;
    }
};
// @lc code=end
